### PR TITLE
Js2 week3

### PR DESCRIPTION
**Pull request by malnaqeeb**

### CODE REVIEW

File: Week3/homework/step2-4.js [hunk 1]
```diff
@@ -1,9 +1,12 @@
...
+// Create a constructor, Dog, with properties name, color, and numLegs that are set to a string, a string, and a number, respectively.
 function Dog() {
-  // add your code here
+  this.name = 'Bo';
+  this.color = 'black and white';
+  this.numLegs = 4;
 }
-
+// Use the Dog constructor above to create a new instance of Dog, assigning it to a variable hound.
 const hound = new Dog();
```
[the remark this answers is not in the document]
Comment (malnaqeeb) [reply Week3/homework/step2-4.js]: ah, now i get what you mean. thanks :+1:

File: Week3/homework/step2-6.js [hunk 2]
```diff
@@ -4,17 +4,21 @@ const arr2d = [[1, 2], [3, 4], [5, 6]];
...
 
 function flattenArray3d(arr) {
-  // Replace this comment and the next line with your code
-  console.log(arr);
+  return arr.flat(2); // depth = 2
+}
+// for unknown number of depths.
+const arrXd = [[1, [2, 2, [3, [4, [5, [6]]]]], 1]];
+function flattenWithXDimension(arr) {
+  return arr.flat(Infinity);
```
[the remark this answers is not in the document]
Comment (malnaqeeb) [reply Week3/homework/step2-6.js]: Thanks :)

File: Week3/homework/step2-3.js [hunk 3]
```diff
@@ -30,9 +32,12 @@ console.log('while', repeatStringNumTimesWithWhile('abc', 3));
 function repeatStringNumTimesWithDoWhile(str, num) {
   // eslint-disable-next-line prefer-const
   let result = '';
-
-  // Replace this comment and the next line with your code
-  console.log(str, num, result);
+  if (num <= 0){ 
+    return result;
```
[the remark this answers is not in the document]
Comment (malnaqeeb) [reply Week3/homework/step2-3.js]: done :+1:

File: Week3/homework/step2-1.js [hunk 4]
```diff
@@ -2,15 +2,15 @@
 
 function foo(func) {
   // What to do here?
-  // Replace this comment and the next line with your code
-  console.log(func);
+  // since we aim to pass bar as argument to foo then foo should return its parameter.
```
[the remark this answers is not in the document]
Comment (malnaqeeb) [reply Week3/homework/step2-1.js]: done :+1:

File: Week3/homework/step2-2.js [hunk 5]
```diff
@@ -2,18 +2,26 @@
...
-  console.log(startIndex, stopIndex, threeCallback, fiveCallback, numbers);
+  for (let elem = startIndex; elem <= stopIndex; elem++) {
+    numbers.push(elem);
+  }
+  numbers.forEach(elem => {
+    if (elem % 3 === 0) {
+      threeCallback(elem);
+    } else if (elem % 5 === 0) {
+      fiveCallback(elem);
+    } else if (elem % 5 === 0 && elem % 3 === 0) {
```
[the remark this answers is not in the document]
Comment (malnaqeeb) [reply Week3/homework/step2-2.js]: done :+1:

File: Week3/homework/step2-2.js [hunk 6]
```diff
@@ -10,9 +10,6 @@ function threeFive(startIndex, stopIndex, threeCallback, fiveCallback) {
       threeCallback(elem);
     } else if (elem % 5 === 0) {
```
Comment (Meno-101) [inline Week3/homework/step2-2.js]: Almost there, with this `else` statement it's stopping right away once it has the `if` condition true, the this else never reached out!
```suggestion
     }
    if (elem % 5 === 0) {
```

Comment (malnaqeeb) [reply Week3/homework/step2-2.js]: Thanks @Meno-101, that was useful :+1:

**Pull request closed without merging**